async-context/README.md

```diff
@@ -0,0 +1,375 @@
+|   Host Continuation B()         |                      |
+===================================            -----------
+```
+
+## Continuation Invocation States
+A `Continuation Invocation` can be in a number of states:
+  - `pending` - A  `Continuation Invocation` instance has been created, but is not yet executing, nor is it ready to execute.
+  - `cancelled` - A `Continuation Invocation` was abandoned.
+  - `ready` - A `Continuation Invocation` is ready to execute, but is not yet executing
+  - `executing` - `Continuation Invocation` is currently executing on the stack
+  - `paused` - A `Continuation Invocation` has started execution, but is not currently on the stack.  E.g., for generator functions. // todo - is this necessary for async/await & generators.  what is tc39 terminology?
+  - `completed` - A `Continuation Invocation` instance is completed.  However, the instance, during execution, may have called other `Continuation Points` and those `Continuation Invocations` are still pending execution.
+  - `collectable` - A `Continuation Invocation` instance is completed and all child-spawned continuations are in the collectable state.
+
+## Context and Current Context
+For any given stack frame, we define it's  **Context** as the `Continuation Invocation` defined by the first `Continuation` below the given frame on the stack.  For example, in the previous diagram, the stack frame of `function x()` has a `Current Context` of `ci3`,
+and the stack frame of  `... host code...` has a "`Current Context` of `ci2`.  We define the **Current Context** as the `Context` of the top frame on the stack. 
+
+## Link Context
+We define the **Link Context** of a `Continuation` to be the `Current Context` when a `Continuation` is constructed.   In the example from above, the `continuation` `f2` has a `link context` pointing to `Continuation Invocation` `ci1`.
+
+## Ready Context
+We define the **Ready Context** to be the `Current Context` when a `Continuation Invocation` changes state from `pending` to `ready`.  This is useful for promises, where we want to understand the series of `Continuation Invocation` instances on a promise chain.  When a promise is resolved or rejected, the promise implementation must modify the state of `Continuation Invocations` associated with subsequent promises on the chain. Consider the following example:
+
+```javascript
+const p = new Promise((resolve, reject) => {
+    setTimeout(function f1() {
+       p.then(function f2() {
+             myLoggingAPI('hello');
+         });
+     }, 1000);
+
+     setTimeout(function f3() {
+         resolve(true);
+     }, 2000)
+});
+```
+
+In this code example, we have:
+  - `setTimeout` and `Promise.then` as our `continuation points`. 
+  - four `continuations` - `f1`, `f2` and `f3`, plus the "root continuation" that 
+    the initial code is executing in. 
+  - four `Continuation Invocations`.
+
+For non-promise-based APIs, the `Ready Context` is the same as the `Link Context`.
```

Comment:
Is this also true for server like objects where functions get called because of some external event (e.g. incoming TCP connection on a net server,...)?

Reply:
I'd like to get to a stronger definition here, and I'm thinking of defining the `X Context` to be the `Current Context` when the `Continuation Invocation` instance changes to state `X`.  I like this because it gives the host some flexibility to account for host-specific things in a way that remains consistent with the definitions.  And, it generalizes the "Ready Context" away from promises. Let me know if you have thoughts on this.